apps/emqx_durable_storage/src/emqx_ds_beamformer.erl

```diff
@@ -173,108 +252,201 @@
     | {ok, end_of_stream}
     | emqx_ds:error(_).
 
--record(s, {
-    module :: module(),
-    metrics_id,
-    shard,
-    name,
-    pending_queue :: ets:tid(),
-    pending_request_limit :: non_neg_integer(),
-    wait_queue :: ets:tid(),
-    is_spinning = false :: boolean(),
-    batch_size :: non_neg_integer()
-}).
+-define(name(SHARD), {n, l, {?MODULE, SHARD}}).
+-define(via(SHARD), {via, gproc, ?name(SHARD)}).
 
--type s() :: #s{}.
+-type gvar() :: #{
+    cbm := module(),
+    fc_tab := ets:tid(),
+    owner_tab := ets:tid(),
+    stuck_sub_tab := sub_tab()
+}.
+
+%% Beam builder:
 
--record(shard_event, {
-    events :: [{_Stream, event_topic()}]
+%% Per-subscription builder state:
+-record(beam_builder_sub, {
+    s :: sub_state(),
+    matcher :: fun(),
+    mask :: emqx_ds_dispatch_mask:enc(),
+    n_msgs :: pos_integer()
 }).
 
--define(fulfill_loop, fulfill_loop).
--define(housekeeping_loop, housekeeping_loop).
+%% Per-node builder:
+-record(beam_builder_node, {
+    global_n_msgs :: non_neg_integer() | undefined,
+    n_msgs = 0 :: non_neg_integer(),
+    %% List of messages and keys, reversed:
+    pack = [] :: [{emqx_ds:message_key(), emqx_types:message()}],
+    subs = #{} :: #{reference() => #beam_builder_sub{}}
```

Comment:
Just curious: is there any difference (e.g., in data size) expected with transposed packing (i.e., mapping messages to the bitmap-encoded set of subscribers)?

Reply:
I am not sure I understand the question.